### ISSUE TITLE
per-parameter gradient clipping if it x% larger than exp_avg

### ISSUE THREAD

**Issue by ibeltagy** [allenai/OLMo#266]
### 🚀 The feature, motivation and pitch

Large spikes start with large gradients that are usually localized in specific parameters. This means global gradient clipping doesn't usually catch it until it is too late and has spread everywhere in the model. 

The proposal here is to employ per-parameter gradient clipping. 

Q1: use `grad.max` or `grad.norm`?
A1: use `grad.norm`. The logs show that either `grad.max` and `grad.norm` spike at the exact same time, or that `grad.norm` spikes a little earlier

Q2: what value of `max_norm` to use for each param?
A2: use adaptive value making `max_norm` a function of `exp_avg`. In other words, clip the param gradient if it is `x`times larger than its historical value (`exp_avg`)

Q3: example of how to implement?
A3: sure. It is probably a single line of code here https://github.com/allenai/LLM/blob/94f120fa8f5c1e46a30b5d8a05dedfa51f6b563c/olmo/optim.py#L149

The code is something like:
```
max_norm = max_norm_ratio * p.get_exp_avg().norm()   # I know `get_exp_avg()` is not a real function, but you know what I mean
```

Q4: why would you use this over update clipping (as in adafactor and stableAdamW)?
A4: because 
- update clipping protects the parameters from the spike but doesn't project `exp_avg` and `exp_avg_sq`, which means both of them will suddenly increase, then a few grad-updates later, update clipping is not effective
- if the spike is because of good thing (say, the model discovered a new "feature" that it really wants to learn), then you want the model to learn this thing not ignore it. If you set `lr=0` for these steps (as in update clipping), the model doesn't learn this new feature, and it keeps discoving it over and over again. Instead, what you want is a smaller gradient in the direction of this good feature that trains the model but does that slowly. 

Q5: Priority?
A5: given that we don't have real spikes now, this is not high priority

Q6: how to develop?
A6: download the logs from wandb, make sure the implementation will trigger a grad-clip for the right params at the right time. Do this before actually training things. 

### Alternatives

This proposal is an alternative to StableAdamW https://github.com/allenai/LLM/pull/252

### Additional context

Easy to implement on top of this PR https://github.com/allenai/LLM/pull/261
